server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -927,12 +899,11 @@ protected final IndexingStrategy planIndexingAsNonPrimary(Index index) throws IO
                 // See testRecoveryWithOutOfOrderDelete for an example of peer recovery
                 plan = IndexingStrategy.processButSkipLucene(false, index.seqNo(), index.version());
             } else {
-                final OpVsLuceneDocStatus opVsLucene = compareOpToLuceneDocBasedOnSeqNo(index);
-                if (opVsLucene == OpVsLuceneDocStatus.OP_STALE_OR_EQUAL) {
-                    plan = IndexingStrategy.processAsStaleOp(softDeleteEnabled, index.seqNo(), index.version());
+                final OptionalLong prevSeqNo = resolveDocSeqNo(index);
+                if (prevSeqNo.isPresent() == false || prevSeqNo.getAsLong() < index.seqNo()) {
+                    plan = IndexingStrategy.processNormally(prevSeqNo.isPresent() == false, index.seqNo(), index.version());
                 } else {
-                    plan = IndexingStrategy.processNormally(opVsLucene == OpVsLuceneDocStatus.LUCENE_DOC_NOT_FOUND,
-                        index.seqNo(), index.version());
+                    plan = IndexingStrategy.processAsStaleOp(softDeleteEnabled, index.seqNo(), index.version(), prevSeqNo.getAsLong());
```

Comment:
can you add an assertion that loads the primary term and checks equality?

Reply:
I added this assertion but it was violated. I added a TODO in resolveDocSeqNo method